scripts/sanitize.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+
+# Copyright 2017 Canonical Ltd.
+# Generate a Javascript file that can be used to sanitize a Mongo database so
+# it can be shared.
+from __future__ import print_function
+
+import sys
+
+# This lists the collections and fields in those collections that need to be sanitized
+to_sanitize = [
+ ('users', ['passwordhash']),
+ ('units', ['passwordhash']),
+ ('machines', ['passwordhash']),
+ ('settings', ['settings']),
+ ('controllers', ['settings', 'cert', 'privatekey', 'caprivatekey', 'sharedsecret', 'systemidentity']),
+ ('actions', ['message', 'results']),
+ ('cloudCredentials', ['attributes']),
+]
+
+def generateScript():
+    lines = []
+    # Create an index on the transactions so that the updates go faster
+    lines.append('print(new Date().toLocaleString())')
+    lines.append('db.txns.createIndex({"o.c": 1})')
+    for collection, attributes in to_sanitize:
+        # First we generate the sanitization of the collection itself
+        # (we don't use .format() because {} is used all the time in Javascript
+        lines.append('print(new Date().toLocaleString())')
+        lines.append('print("updating collection %s for %s")' % (collection, attributes))
+        lines.append('print(db.%s.update({}, {"$set": {' % (collection,))
+        for attribute in attributes:
+            lines.append('    "{}": "REDACTED",'.format(attribute))
+        # replace the last ',' with a '}'
+        lines[-1] = lines[-1][:-1] + '}'
+        lines.append('}, {"multi": 1}))')
+
+        # Now update the TXN records for insert and update
+        for attribute in attributes:
+            lines.append('print(new Date().toLocaleString())')
+            lines.append('print("updating insert txns for %s %s")' % (collection, attribute))
+            lines.append('print(db.txns.update({"o.c": "%s", "o.i.%s": {"$exists": 1}}, {"$set": {' % (collection, attribute))
+            lines.append('    "o.$.i.%s": "REDACTED"}' % (attribute,))
+            lines.append('}, {"multi": 1}))')
+            lines.append('print(new Date().toLocaleString())')
+            lines.append('print("updating update txns for %s %s")' % (collection, attribute))
+            # Our TXN entries have a '$set' with a literal $ in them.
+            # Apparently Mongo is perfectly fine for you to create documents
+            # and query documents with $, but won't let you "update" documents
+            # with a $ in them. so we just unset those fields instead of setting them to 'REDACTED'.
+            lines.append('print(db.txns.update({"o.c": "%s", "o.u.$set.%s": {"$exists": 1}}, {"$unset": {' % (collection, attribute))
+            lines.append('    "o.$.u.$set.%s": "1"}' % (attribute,))
+            lines.append('}, {"multi": 1}))')
+        lines.append('')
+    lines.append('db.txns.deleteIndex({"o.c": 1})')
+    lines.append('print(new Date().toLocaleString())')
+    return lines
```

Comment:
With the change I suggested above, there's actually no need to build a list. Just `yield` the lines instead.